finagle-core/src/test/scala/com/twitter/finagle/filter/RequestMeterFilterTest.scala

```diff
@@ -0,0 +1,44 @@
+package com.twitter.finagle.filter
+
+import com.twitter.concurrent.AsyncMeter
+import com.twitter.conversions.time._
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.finagle.{Failure, Service}
+import com.twitter.util.{Await, Future}
+import org.junit.runner.RunWith
+import org.scalatest.FunSuite
+import org.scalatest.junit.JUnitRunner
+
+@RunWith(classOf[JUnitRunner])
+class RequestMeterFilterTest extends FunSuite {
+
+  val timer = DefaultTimer.twitter
+  val meter = AsyncMeter.newMeter(1, 1 second, 1)(timer)
+
+  test("return service execution after getting a permit") {
+    val echoSvc = new Service[Int, Int] {
+      def apply(req: Int) = Future(req)
+    }
+    val svc = new RequestMeterFilter(meter) andThen echoSvc
+    assert(Await.result(svc(2)) == 2)
+  }
+
+  test("mark dropped requests as rejected") {
+    val neverSvc = new Service[Int, Int] {
+      def apply(req: Int) = Future.never
+    }
+    val svc = new RequestMeterFilter(meter, 2) andThen neverSvc
+    val f = intercept[Failure] { Await.result(svc(1)) }
+    assert(f.isFlagged(Failure.Restartable))
+  }
+
+  test("service failures are not wrapped as rejected") {
+    val exc = new Exception("app exc")
```

Comment:
So the way that AsyncMeter works is that it allows `burst` through at a time–if you examine the first future, it has already been satisfied.  So only the second one is waiting, which means that there is only one waiter.  It would be incorrect for it to throw.  If instead you do:

```scala
val timer = new MockTimer
Time.withCurrentTimeFrozen { ctl =>
  val meter = AsyncMeter.newMeter(1, 1 second, 1)(timer)
  val f1 = meter.await(1)
  assert(f1.isDone)
  val f2 = meter.await(1)
  assert(!f2.isDefined)
  val f3 = meter.await(1)
  assert(f3.isDefined)
  intercept[RejectedExecutionException] { Await.result(f3, 5.seconds) }
  ctl.advance(1.second)
  timer.tick()
  assert(f2.isDone)
}
```

In fact, we have tests like this for [AsyncMeter][0] already.

[0]: https://github.com/twitter/util/blob/develop/util-core/src/test/scala/com/twitter/concurrent/AsyncMeterTest.scala#L53